packages/notebook/src/searchprovider.ts

```diff
@@ -0,0 +1,571 @@
+   * @param translator Application translator
+   */
+  constructor(
+    widget: NotebookPanel,
+    protected translator: ITranslator = nullTranslator
+  ) {
+    super(widget);
+
+    this.widget.model!.cells.changed.connect(this._onCellsChanged, this);
+    this.widget.content.activeCellChanged.connect(
+      this._onActiveCellChanged,
+      this
+    );
+    this.widget.content.placeholderCellRendered.connect(
+      this._onPlaceholderRendered,
+      this
+    );
+  }
+
+  /**
+   * Report whether or not this provider has the ability to search on the given object
+   *
+   * @param domain Widget to test
+   * @returns Search ability
+   */
+  static isApplicable(domain: Widget): domain is NotebookPanel {
+    // check to see if the CMSearchProvider can search on the
+    // first cell, false indicates another editor is present
+    return domain instanceof NotebookPanel;
+  }
+
+  /**
+   * Instantiate a search provider for the notebook panel.
+   *
+   * #### Notes
+   * The widget provided is always checked using `isApplicable` before calling
+   * this factory.
+   *
+   * @param widget The widget to search on
+   * @param translator [optional] The translator object
+   *
+   * @returns The search provider on the notebook panel
+   */
+  static createNew(
+    widget: NotebookPanel,
+    translator?: ITranslator
+  ): ISearchProvider {
+    return new NotebookSearchProvider(widget, translator);
+  }
+
+  /**
+   * The current index of the selected match.
+   */
+  get currentMatchIndex(): number | null {
+    let agg = 0;
+    let found = false;
+    for (let idx = 0; idx < this._searchProviders.length; idx++) {
+      const provider = this._searchProviders[idx];
+      const localMatch = provider.currentMatchIndex;
+      if (localMatch !== null) {
+        agg += localMatch;
+        found = true;
+        break;
+      } else {
+        agg += provider.matchesCount;
+      }
+    }
+    return found ? agg : null;
+  }
+
+  /**
+   * The number of matches.
+   */
+  get matchesCount(): number | null {
+    return this._searchProviders.reduce(
+      (sum, provider) => (sum += provider.matchesCount),
+      0
+    );
+  }
+
+  /**
+   * Set to true if the widget under search is read-only, false
+   * if it is editable.  Will be used to determine whether to show
+   * the replace option.
+   */
+  get isReadOnly(): boolean {
+    return this.widget?.content.model?.readOnly ?? false;
+  }
+
+  /**
+   * Dispose of the resources held by the search provider.
+   *
+   * #### Notes
+   * If the object's `dispose` method is called more than once, all
+   * calls made after the first will be a no-op.
+   *
+   * #### Undefined Behavior
+   * It is undefined behavior to use any functionality of the object
+   * after it has been disposed unless otherwise explicitly noted.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    super.dispose();
+
+    this.widget.content.placeholderCellRendered.disconnect(
```

Comment:
Same question re: clearing signal data:
```suggestion
    Signal.clearData(this);
```